### ISSUE TITLE
[no-this-assignment] unfixable reporting

### ISSUE THREAD

**Issue by JounQin** [sindresorhus/eslint-plugin-unicorn#1108]
```ts
class A {
  @Input()
  input: boolean

  test() {
    const self = this // error1
    let currentSeries: Highcharts.Series
    return {
      mouseOver() {
        currentSeries = this // error2
        if (!self.input) {
          return
        }
        // ...
      },
      mouseOut() {
        console.log(currentSeries)
      },
    }
  }
}
```

I don't think it can be implemented by arrow function or bind.

**Comment by fisker** [comment]
```js
class A {
  test() {
    return {
      mouseOver: (series) => {
        currentSeries = series
        if (!this.input) {
          return
        }
     }
    }
  }
}

const a = new A();

Highcharts.chart({
 ///
  mouseOver() {
    return a.mouseOver(this)
  }
}}
```


Too much?

**Comment by JounQin** [comment]
I don't think we should change the code logic...

**Comment by fisker** [comment]
`eslint-disable`

**Comment by JounQin** [comment]
> eslint-disable

Well, that's what I'm doing temporarily, but it would be better to fix it in this rule.

**Comment by fisker** [comment]
I don't think there is nothing we can do, this rule intentionally require some refactor, see last example in docs

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-this-assignment.md#pass

**Comment by JounQin** [comment]
I understand that `this rule intentionally require some refactor`, but I don't think your proposed refactor is acceptable. It changes the code logic too much and may cause unexpected bugs, and also reduce the readability.

The point is that we need to access the internal `this` and outside `this` at the same time.

Of course, I understand even if you decide `won't fix`.

**Comment by fisker** [comment]
Does Highcharts Event have a `.target` like `Event` in DOM?

**Comment by JounQin** [comment]
We can't get `this` via the native `event` param (I've tried before to access `this` by this way before I use this pattern currently). And I don't think we should consider `Highcharts` specifically. It's just code.

**Comment by JounQin** [comment]
Options like [`typescript-eslint/no-this-alias`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-this-alias.md#options) would help maybe.

**Comment by fregante** [comment]
The rule is about preventing `self`, so it's working as intended here, even if the fix isn't strictly possible without a large refactor. Such is life with lint rules
